<a href="https://colab.research.google.com/github/TreborRP30/An-lisis-inmobiliario-de-lima-2019/blob/main/01_Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Preparación de la mesa de trabajo**

En esta etapa se hace lo básico que es importar todas las librerías necesarias para esta limpieza de la base de datos.

In [4]:
#Importación masiva de librerías esenciales
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

#Configuración visual para tus gráficos (para que se vean profesionales)
sns.set_theme(style="whitegrid")
plt.rcParams['figure.figsize'] = (10, 6)

In [5]:
#lectura
ruta = "/content/drive/MyDrive/urbania_data (1).csv"
df = pd.read_csv(ruta)
print(df.shape)
df.head()

(7663, 96)


,Unnamed: 0.1,Unnamed: 0,Antiguedad,Anunciante,Balneario,NroBanios,Nro_pisos,Cocheras,Descripcion,Direccion,...,'Aire acondicionado','Amoblado','Sauna','Calefaccion','Equipado','Terma','Portero electrico','Cocina con reposteros',Distrito,Provincia
0,0,3,5.0,Valentinoinmobiliaria,Nulo,4.0,2.0,3.0,excelente opcion para vivir en una de las mej...,"Malecon Monte Bello 5 La Molina, La Planicie, ...",...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,LaMolina,Lima
1,1,4,5.0,No disponible,Nulo,5.0,3.0,1.0,hermosa casa a espalda de futura avenida prol...,Mz. O2 Lt. 03 Urb. San Antonio De Carapongo 2 ...,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,Lurigancho,Lima
2,2,5,3.0,No disponible,Nulo,5.0,3.0,4.0,casa de estreno acabados de primera. calle ce...,"El Pinar 120 , La Planicie 1era Etapa La Molin...",...,1.0,1.0,1.0,0.0,1.0,1.0,1.0,1.0,LaMolina,Lima
3,5,8,1.0,No disponible,Sol Y Mar,5.0,2.0,3.0,linda casa en condominio con piscina club hou...,"Km 98.5 De La Panamericana Sur Asia, Sol Y Mar...",...,0.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,Asia,Lima
4,6,9,4.0,No disponible,Nulo,4.0,2.0,2.0,monterrico 2 casas en 1 ( casa bajos 4 dormit...,"Mariano De Rivera Y Ustariz #215 Surco, Huerto...",...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,SantiagoDeSurco,Lima


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7663 entries, 0 to 7662
Data columns (total 96 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Unnamed: 0.1                    7663 non-null   int64  
 1   Unnamed: 0                      7663 non-null   int64  
 2   Antiguedad                      7663 non-null   float64
 3   Anunciante                      7663 non-null   object 
 4   Balneario                       7663 non-null   object 
 5   NroBanios                       7663 non-null   float64
 6   Nro_pisos                       7663 non-null   float64
 7   Cocheras                        7663 non-null   float64
 8   Descripcion                     7663 non-null   object 
 9   Direccion                       7663 non-null   object 
 10  Dormitorios                     7663 non-null   object 
 11  Estado de Inmueble              7663 non-null   object 
 12  Fecha_pub                       76

Realizó un análisis exploratorio de mi dataset para saber su estado, y lo que tomo en cuenta es:

•La cantidad de columnas y filas: 7663 columnas y 96 filas.

•Los tipos de datos de las columnas: 8 decimales, 3 enteros y 85 categoricas.

•Si hay celdas vacias o valor nulos, en mi caso no hay.

Igualmente eliminare todos las filas con un dato faltantes o nulo, en mi caso se puede permitir por el tamaño de mi dataset.

In [7]:
df.dropna(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7663 entries, 0 to 7662
Data columns (total 96 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Unnamed: 0.1                    7663 non-null   int64  
 1   Unnamed: 0                      7663 non-null   int64  
 2   Antiguedad                      7663 non-null   float64
 3   Anunciante                      7663 non-null   object 
 4   Balneario                       7663 non-null   object 
 5   NroBanios                       7663 non-null   float64
 6   Nro_pisos                       7663 non-null   float64
 7   Cocheras                        7663 non-null   float64
 8   Descripcion                     7663 non-null   object 
 9   Direccion                       7663 non-null   object 
 10  Dormitorios                     7663 non-null   object 
 11  Estado de Inmueble              7663 non-null   object 
 12  Fecha_pub                       76

🛠️ **Limpieza Estructural y Estandarización del Dataset**

En esta fase de limpieza inicial, se aplicaron transformaciones directas a la estructura del DataFrame para optimizar la calidad de los datos y preparar la información para el análisis exploratorio y el modelado predictivo. El proceso consistió en tres acciones clave:
1. Eliminación de variables no predictivas (Reducción de dimensionalidad): Se descartaron columnas que representan "ruido" analítico. Esto incluye identificadores internos del scraping web (match), datos comerciales sin impacto directo en el valor del inmueble (Anunciante), texto libre no estructurado que requiere procesamiento avanzado (Descripcion, Direccion) y variables métricas duplicadas (Area_constr, Area_total).
¿Por qué se hizo? Eliminar estas columnas reduce el peso del dataset en memoria, evita la redundancia de información y asegura que los futuros algoritmos se enfoquen únicamente en variables que realmente influyen en el precio.
2. Corrección de errores de codificación: Se aplicó un mapeo estructurado para renombrar columnas que presentaban caracteres corruptos o errores tipográficos generados durante la extracción de los datos (por ejemplo, transformar 'Jardanxadn' a 'Jardin').
¿Por qué se hizo? Garantiza la integridad y legibilidad de los datos, asegurando que las variables sean comprensibles al momento de generar gráficos o presentar resultados.
3. Estandarización de la sintaxis (Buenas prácticas de Python): Se procesaron todos los nombres de las columnas restantes para eliminar caracteres especiales (como comillas simples) y reemplazar los espacios en blanco por guiones bajos (_).
¿Por qué se hizo? Es una convención técnica estándar en la programación con Pandas. Permite acceder a las columnas utilizando la notación de punto (ej. df.Cuarto_de_servicio en lugar de requerir corchetes y comillas df['Cuarto de servicio']), lo que agiliza la escritura del código y previene errores de sintaxis en etapas posteriores.

In [11]:
# 1. ELIMINAR columnas basura
columnas_basura = ['match', 'Anunciante', 'Direccion', 'Descripcion', 'Area_constr', 'Area_total','Unnamed:_0','Unnamed:_0.1']
df.drop(columns=columnas_basura, inplace=True, errors='ignore')

# 2. RENOMBRAR las mal tipeadas
renombres = {
    'anx81tico': 'Atico',
    'Jardanxadn_Interno': 'Jardin_interno',
    'Lavanderanxada': 'Lavanderia',
    'Guardiananxada': 'Guardiania',
    'Guarderanxada': 'Guarderia',
    'Jardanxadn_Interno': 'Jardin_interno',
    'anx81reas_verdes': 'Areas_verdes',
    'Areadeportiva': 'Area_deportiva',
    'Areade_BBQ': 'Area_BBQ',
    'Areade_sauna': 'Area_sauna',
    'Desagaue': 'Desague',
    'NroBanios': 'Nro_banios',
    'Juegos_para_ninios': 'Juegos_ninos'
}
df.rename(columns=renombres, inplace=True)

# 3. Formatear todas las columnas restantes (quitar espacios y comillas sueltas)
# Esto convierte "Cuarto de servicio" en "Cuarto_de_servicio" automáticamente
df.columns = df.columns.str.replace("'", "").str.replace(" ", "_")

# Verificamos el resultado
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7663 entries, 0 to 7662
Data columns (total 88 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Antiguedad                    7663 non-null   float64
 1   Balneario                     7663 non-null   object 
 2   Nro_banios                    7663 non-null   float64
 3   Nro_pisos                     7663 non-null   float64
 4   Cocheras                      7663 non-null   float64
 5   Dormitorios                   7663 non-null   object 
 6   Estado_de_Inmueble            7663 non-null   object 
 7   Fecha_pub                     7663 non-null   object 
 8   Luminosidad                   7663 non-null   object 
 9   Mascotas                      7663 non-null   object 
 10  Precio                        7663 non-null   float64
 11  Tipo                          7663 non-null   object 
 12  TipoCochera                   7663 non-null   object 
 13  Ubi

In [18]:
#Limpio caracteres comunes si los hubiera (comas, letras, símbolos)
#(Si la columna de precio tiene comas, las quito primero)
df['Precio'] = df['Precio'].astype(str).str.replace(',', '', regex=True)

#(Reemplaza 'Area_total_m2' por el nombre real de tu columna. Si tiene "m2", lo quito)
df['Area_total_m2'] = df['Area_total_m2'].astype(str).str.replace('m2', '', regex=True)

#Fuerzo la conversión a numérico.
#El parámetro errors='coerce' es clave: si encuentra un texto que no puede convertir (como un valor "Sin dato"), lo vuelve NaN (nulo) para no romper el código.
df['Precio'] = pd.to_numeric(df['Precio'], errors='coerce')
df['Area_total_m2'] = pd.to_numeric(df['Area_total_m2'], errors='coerce') # Cambia 'Area_total_m2' por tu columnas

📊 **Análisis Estadístico Descriptivo y Detección de Outliers**

En esta fase exploratoria, generamos un resumen estadístico descriptivo utilizando el método .describe() sobre las variables numéricas más críticas para el análisis (Precio, Area_constr_m2, Nro_banios). El objetivo principal de esta operación es auditar la integridad matemática de los datos para comprender su distribución inicial e identificar valores atípicos (outliers) antes de iniciar la etapa de modelado.

Hallazgos clave del diagnóstico:

Identificación de Extremos (Mínimos y Máximos): El análisis de los indicadores min y max expuso anomalías severas que comprometen la calidad del dataset. Como evidencia, la variable Area_constr_m2 arrojó valores mínimos físicamente imposibles para una vivienda (1.0 m²) y máximos desproporcionados (8,000.0 m²). Esto confirma la presencia de ruido originado por errores de digitación al momento de publicar los anuncios web.

Evaluación de la Dispersión:

Al comparar la media (mean) con la mediana (el percentil 50%), se observan distorsiones provocadas por estos valores atípicos. Las propiedades con precios o áreas absurdamente altas están "jalando" el promedio hacia arriba, alejándolo del comportamiento real del mercado (representado entre los cuartiles 25% y 75%).

In [ ]:
#Definimos las variables numéricas más críticas de tu dataset
columnas_clave = ['Precio', 'Area_total_m2', 'Area_constr_m2', 'Dormitorios', 'Nro_banios']

#Generamos un resumen estadístico, mostrando los valores mínimos y máximos
df[columnas_clave].describe().round(2)

In [ ]:
#Guardamos el total de filas iniciales
filas_antes = df.shape[0]

#Filtramos usando sentido común inmobiliario
df_limpio = df[
    (df['Area_constr_m2'] >= 15) &    # Un cuarto/minidepa muy pequeño mide al menos 15m2
    (df['Area_constr_m2'] <= 500) &   # Más de 500m2 construidos ya es atípico para vivienda regular
    (df['Precio'] >= 25000)           # Ponemos un límite mínimo razonable para el precio
]

filas_despues = df_limpio.shape[0]
print(f"Limpieza lista. Se eliminaron {filas_antes - filas_despues} filas con errores extremos.")
print(f"Te quedan {filas_despues} propiedades listas para analizar.")

#Reemplazamos el dataframe original con el limpio
df = df_limpio.copy()

📉 **Análisis de Varianza y Reducción de Dimensionalidad**

En esta fase de limpieza, el enfoque pasó de las variables numéricas a las variables categóricas o booleanas (características binarias como la presencia de Sauna, Sala_de_cine, etc.). En este tipo de datos no se buscan "valores extremos", sino que se audita el desbalance de clases o la baja varianza.

Acciones técnicas realizadas:

Auditoría Automática de Varianza: Se implementó un algoritmo iterativo para escanear todas las columnas del dataset e identificar aquellas que presentaban un comportamiento "casi constante". Se estableció un umbral estadístico estricto: si un único valor dominaba más del 95% de los registros en una columna (por ejemplo, si el 98% de los inmuebles "No" tiene Sala_de_cine), la variable fue marcada para revisión.

Depuración por Carencia de Valor Predictivo: Las características identificadas con nula diversidad fueron eliminadas estratégicamente del DataFrame. ¿Por qué se hizo? En el modelado de datos y Machine Learning, si una variable no fluctúa, no aporta información útil para que el algoritmo encuentre patrones.

Mantener columnas casi idénticas en todas las filas solo aumenta el ruido, consume memoria innecesaria y provoca la "maldición de la dimensionalidad" sin ayudar a predecir el precio del inmueble.

In [ ]:
#Auditamos de Baja Varianza (Desbalance)
print("🔍 Analizando columnas casi constantes (más del 95% iguales):\n")

columnas_baja_varianza = []

#Iteramos por todas las columnas del dataset
for col in df.columns:
    # Aseguramos que la columna tenga datos
    if df[col].count() > 0:
        # Calculamos el porcentaje del valor que más se repite
        freq_maxima = df[col].value_counts(normalize=True).iloc[0] * 100

        #Si un solo valor domina más del 95% de la columna...
        if freq_maxima > 95:
            print(f"🚨 {col}: {freq_maxima:.2f}% de los datos son idénticos.")
            columnas_baja_varianza.append(col)

print(f"\nTotal de variables que no aportan valor predictivo: {len(columnas_baja_varianza)}")

# 2. Eliminación de las columnas sin varianza
df.drop(columns=columnas_baja_varianza, inplace=True, errors='ignore')

print(f"Limpieza completada. Ahora te quedan {df.shape[1]} columnas y {df.shape[0]} filas útiles en tu dataset.")

📋 **Auditoría Final de Calidad de Datos (Quality Assurance)**

Antes de dar por concluida la fase de limpieza estructural, se ejecutó un control de calidad (QA Check) de tres niveles para garantizar la viabilidad del dataset frente a algoritmos de Machine Learning. Esta auditoría sistemática es un estándar en la industria para prevenir errores silenciosos en la fase de modelado.

Controles ejecutados:

Verificación de Registros Duplicados: Se escaneó el dataset en busca de filas idénticas. En proyectos de web scraping, es habitual extraer el mismo anuncio inmobiliario múltiples veces debido a la paginación dinámica de los sitios web. Eliminar estos duplicados evita el sobreajuste (overfitting) y asegura que las métricas de promedio no estén sesgadas.

Auditoría de Valores Nulos Residuales (NaN): Se realizó un mapeo de las columnas clave para asegurar que, tras las transformaciones y eliminaciones previas, no quedaran campos vacíos ocultos que pudieran interrumpir la ejecución matemática de futuros modelos.

Validación de Tipos de Datos (Dtypes): Se revisó la estructura interna de las variables predictoras (Precio, Area, Dormitorios, etc.). El objetivo de este paso crítico es asegurar que el motor de Pandas reconozca las variables matemáticas como formatos numéricos (float64 o int64) y no como cadenas de texto (object). Ignorar este paso resultaría en fallos de compilación al intentar calcular distancias estadísticas o coeficientes de regresión.

In [ ]:
print("--- 1. VERIFICACIÓN DE DUPLICADOS ---")
duplicados = df.duplicated().sum()
print(f"Tienes {duplicados} filas exactamente iguales.")
# Si hay duplicados, los eliminamos
if duplicados > 0:
    df.drop_duplicates(inplace=True)
    print("¡Duplicados eliminados!")

print("\n--- 2. VERIFICACIÓN DE NULOS (TOP 10) ---")
# Esto te mostrará las 10 columnas con más datos faltantes
nulos = df.isnull().sum().sort_values(ascending=False).head(10)
print(nulos[nulos > 0]) # Solo muestra si hay mayores a 0

print("\n--- 3. VERIFICACIÓN DE TIPOS DE DATOS (Dtypes) ---")
# Verificamos si Pandas reconoció bien los números
print("Revisa si alguna de estas numéricas aparece como 'object':")
# Muestra el tipo de dato de nuestras columnas clave
print(df[['Precio', 'Area_constr_m2', 'Nro_banios', 'Dormitorios']].dtypes)

🛠️ **Corrección de Tipos de Datos (Casting) y Formateo Numérico**

Durante la auditoría final de calidad de datos (Dtypes), se detectó que la variable clave Dormitorios estaba catalogada como object (cadena de texto) en lugar de un formato numérico.

Esto suele ser provocado por caracteres alfanuméricos residuales del scraping web.

Para solucionarlo, se aplicó una técnica de coerción numérica mediante la función pd.to_numeric(errors='coerce'). Este método fuerza la transformación de la columna a formato de coma flotante (float64), convirtiendo automáticamente cualquier cadena de texto no parseable en valores nulos (NaN), los cuales fueron posteriormente removidos. Esto garantiza la compatibilidad total del dataset con futuros algoritmos de Machine Learning.

In [ ]:
#Vemos qué valores únicos raros hay antes de la limpieza (para entender el error)
print("Algunos valores actuales en la columna:", df['Dormitorios'].unique()[:10])

#Forzamos la conversión a número.
#errors='coerce' convierte cualquier texto rebelde (como "Sin dato") en nulo (NaN)
df['Dormitorios'] = pd.to_numeric(df['Dormitorios'], errors='coerce')

#Verificamos si la conversión generó valores nulos y los eliminamos
nulos_generados = df['Dormitorios'].isnull().sum()
if nulos_generados > 0:
    print(f"\nSe encontraron {nulos_generados} textos que no eran números. Eliminando esas filas...")
    df.dropna(subset=['Dormitorios'], inplace=True)

#Verificación final del tipo de dato
print("\n--- NUEVOS TIPOS DE DATOS ---")
print(df[['Precio', 'Area_constr_m2', 'Nro_banios', 'Dormitorios']].dtypes)

#Imprimimos el dataset ya limpio y listo para su análisis
print(f"Limpieza completada. Ahora te quedan {df.shape[1]} columnas y {df.shape[0]} filas útiles en tu dataset.")

Y para terminar creamos una copia de la Base de Datos para tener siempre un respaldo de todo nuestro Datos

In [20]:
# Definimos la ruta en tu Drive, pero le cambiamos el nombre al archivo
# para no sobrescribir el original (¡siempre conserva tu raw data!)
ruta_limpia = "/content/drive/MyDrive/urbania_data_LIMPIO.csv"

# Guardamos el dataset en formato CSV.
# 'index=False' evita que Pandas guarde la columna de numeración de las filas, que es ruido.
df.to_csv(ruta_limpia, index=False)

print("¡Éxito! Tu dataset limpio ha sido guardado en tu Drive.")

¡Éxito! Tu dataset limpio ha sido guardado en tu Drive.
